### PR TITLE
Create new NVM chip (25LC1024) library.

### PR DESCRIPTION
**Pull request by mass**
Let me know your thoughts on the code. I can change the interface if you think we should do things in a different way. The high-level API (with fd's) seems to be working well. I think it will be the easiest thing for people to use to store and fetch structs in NVM across reboots.

I recommend you squash merge this PR when ready. I think if we move to a model where we do major work in feature branches, and then squash merge them back into master, it will keep things really clean. But it's up to you how you want to do it.

### CODE REVIEW

File: FSAE.X/FSAE_nvm.c [hunk 1]
```diff
@@ -4,107 +4,342 @@
...
  * Compiler:  Microchip XC32
  * Author:    Andrew Mass
- * Created:   2015-2016
+ * Created:   2017-2018
  */
 #include "FSAE_nvm.h"
 
-SPIConn nvmConnections[10];
-uint8_t nvmConnIdx = 0;
+#define NUM_CONN 10
```
[the remark this answers is not in the document]
Comment (mass) [reply FSAE.X/FSAE_nvm.c]: Don't see any reason to, it's not needed by other files and would only pollute the "namespace"

File: FSAE.X/FSAE_nvm.c [hunk 2]
```diff
@@ -4,107 +4,342 @@
...
+
+_NvmSuperblock superblocks[NUM_CONN];
+
+uint8_t idx(SPIConn* conn) {
+  return ((conn - connections) / sizeof(SPIConn));
+}
+
+uint16_t num_pages(uint32_t size) {
+  // Use ceil() so that blocks always start at a page boundary
+  uint16_t pages = size / 256;
```
[the remark this answers is not in the document]
Comment (mass) [reply FSAE.X/FSAE_nvm.c]: It's possible, but there are a lot of places in this library that are hardcoded for things that would almost definitely change if we moved to a different NVM chip. I think it's pretty hard to make a low-level library that is generalized to a variety of NVM chips. The problem with a page size define is that there are other things that rely on it being 256, like the mask 0xFF and certain variables being uint8_t. I suppose we could #define all that but I'm not sure it's worth it

File: FSAE.X/FSAE_nvm.c [hunk 3]
```diff
@@ -4,107 +4,342 @@
...
  */
-void nvm_write_enable(SPIConn *conn) {
-  spi_send(IN_WREN, conn);
+uint8_t nvm_alloc(SPIConn* conn, uint8_t* block_id, uint32_t vcode,
+                  uint32_t size, uint8_t* fd) {
+  uint16_t i;
+  _NvmSuperblock* sb = &(superblocks[idx(conn)]);
+
+  // Search for matching node
+  for (i = 0; i < 256; i++) {
```
[the remark this answers is not in the document]
Comment (mass) [reply FSAE.X/FSAE_nvm.c]: This too is 256 because the `fd` is a uint8_t. I think we would only need to generalize it if we wanted less nodes than 256

**Comment by mass** [comment]
@jake1317 what's the status of this PR?

**Comment by jake1317** [comment]
The status is I should have merged it and I haven't :) I'll do it now.

**Pull request merged**